### ISSUE TITLE
Bound TypeVar as Self type doesn't behave like a normal TypeVar

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#3913]
**Describe the bug**
Using a bound TypeVar as Self type doesn't behave like a normal TypeVar does, it appears to be widened unnecessarily. 

**Screenshots or Code**
```python
from typing import TypeVar

ST = TypeVar("ST", bound=S)

class S:
    def r(self: ST) -> ST: ...

class Sub(S): ...

reveal_type(S.r(Sub()))  # type is S **not** Sub as expected
```

**Expected behavior**
mypy has what I believe to be the correct result here https://mypy-play.net/?mypy=latest&python=3.10&gist=5207e15b920ead38cc925805f87ba0bc

**VS Code extension or command-line**
pylance v2022.8.50

**Additional context**
The same issue happens with the Self type

**Comment by erictraut** [comment]
I agree that mypy's behavior here is more intuitive, but there's a problem with mypy's approach. It affects generic classes. Consider the case where the base class `S` is generic.

```python
from typing import Generic, TypeVar

ST = TypeVar("ST", bound="S")
T = TypeVar("T")

class S(Generic[T]):
    def r(self: ST) -> ST:
        ...

class Sub(S[int]):
    ...

reveal_type(S[str].r)
reveal_type(S[str].r(Sub())) # This should generate an error, but mypy does not
```

The problem is that mypy isn't applying the specialized type of `S` to `ST` when accessing `S[str].r`. It's leaving the type variable `ST` unsolved. Pyright currently binds `S[str]` to `ST` when evaluating the expression `S[str].r`. (In your example above, it's binding `S` to `ST`, which is why you're seeing the behavior observed in your bug report.)

I'll need to give more thought to the implications of matching mypy's behavior here.

**Comment by erictraut** [comment]
After thinking about this more, I think that pyright's current behavior is correct. The expression `A.b` should establish the type of the `self` parameter for method `b`. As explained above, is especially important if `A` is generic.

**Comment by Gobot1234** [comment]
This differs from runtime python behaviour IMO
```py
In [2]: class S:
   ...:     def r(self):
   ...:         ...
   ...: 
   ...: class Sub(S):
   ...:     ...
   ...: 

In [3]: Sub().r.__self__
Out[3]: <__main__.Sub at 0x10544f7f0>
```
Would it be at all possible to allow the original code to pass but disallow the specialisation issues? I can't actually use Sub.r(Sub()) in my code so currently, I have multiple `# type: ignores` in the same method.

```py
    def parse(self, data: bytes, msg: int = MISSING) -> Self:
        if msg is MISSING:  # case CMsgMulti().parse(data)
            return betterproto.Message.parse(self, data)

        self.header.parse(data)
        if msg in SERVICE_EMSGS:
            try:
                new_class = UMS[self.header.job_name_target][msg in RESPONSE_EMSGS]
            except KeyError:
                log.debug(f"Received an unknown UM {self.header.job_name_target} (%s)", data)
                return self
        else:
            try:
                new_class: type[Self] = PROTOBUFS[msg]  # type: ignore  # save the extra lookup when casting to EMsg
            except KeyError:
                log.debug(f"Received an unknown {EMsg(msg)!r} (%s)", data)
                return self

        self.__class__ = new_class
        return betterproto.Message.parse(self, data[self.header.length :])

**Comment by erictraut** [comment]
I think pyright's current behavior is correct, so I don't plan to make any changes here.
